libraries/c_sdk/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -416,92 +406,106 @@ void _IotMqtt_InvokeSubscriptionCallback( _mqttConnection_t * pMqttConnection,
-                                                     pCurrentLink,
-                                                     _topicMatch,
-                                                     &topicMatchParams );
+     * the array head. */
 
-        /* No subscription found. Exit loop. */
-        if( pCurrentLink == NULL )
-        {
-            break;
-        }
-        else
+    while( index < MAX_NO_OF_MQTT_SUBSCRIPTIONS )
+    {
+        if( contextIndex >= 0 )
         {
-            EMPTY_ELSE_MARKER;
-        }
-
-        /* Subscription found. Calculate pointer to subscription object. */
-        pSubscription = IotLink_Container( _mqttSubscription_t, pCurrentLink, link );
+            index = IotMqtt_FindFirstMatch( &( connToContext[ contextIndex ].subscriptionArray ),
+                                            index,
+                                            &topicMatchParams );
 
-        /* Subscription validation should not have allowed a NULL callback function. */
-        IotMqtt_Assert( pSubscription->callback.function != NULL );
+            /* No subscription found. Exit loop. */
+            if( index == -1 )
+            {
+                break;
+            }
+            else
+            {
+                pSubscription = &( connToContext[ contextIndex ].subscriptionArray[ index ] );
+            }
 
-        /* Increment the subscription's reference count. */
-        ( pSubscription->references )++;
+            /* Subscription validation should not have allowed a NULL callback function. */
+            IotMqtt_Assert( pSubscription->callback.function != NULL );
 
-        /* Copy the necessary members of the subscription before releasing the
-         * subscription list mutex. */
-        pCallbackContext = pSubscription->callback.pCallbackContext;
-        callbackFunction = pSubscription->callback.function;
+            /* Increment the subscription's reference count. */
+            ( pSubscription->references )++;
 
-        /* Unlock the subscription list mutex. */
-        IotMutex_Unlock( &( pMqttConnection->subscriptionMutex ) );
+            /* Copy the necessary members of the subscription before releasing the
+             * subscription list mutex. */
+            pCallbackContext = pSubscription->callback.pCallbackContext;
+            callbackFunction = pSubscription->callback.function;
 
-        /* Set the members of the callback parameter. */
-        pCallbackParam->mqttConnection = pMqttConnection;
-        pCallbackParam->u.message.pTopicFilter = pSubscription->pTopicFilter;
-        pCallbackParam->u.message.topicFilterLength = pSubscription->topicFilterLength;
+            /* Unlock the subscription array mutex. */
+            if( IotMutex_Give( &( connToContext[ contextIndex ].subscriptionMutex ) ) == false )
+            {
+                /* Fail to give subscription mutex as no space is available on queue. */
+                mutexStatus = false;
```

Comment:
Similar comment for updating the mutex status as above

Reply:
Addressed